### PR TITLE
[NETBEANS-1622] Sort editor tabs when a document opens

### PR DESCRIPTION
**Pull request by junichi11**
https://issues.apache.org/jira/browse/NETBEANS-1622

Sort tabs
- By file name
- By File name with parent directory name
- By full file path

**NOTE:** only when a document opens
nb-sort-editor-tabs

### CODE REVIEW

**Comment by geertjanw** [comment]
Pretty cool. :-)

File: platform/core.windows/src/org/netbeans/core/windows/model/DefaultModeModel.java [hunk 1]
```diff
@@ -142,6 +146,10 @@ public void insertOpenedTopComponent(TopComponent tc, int index) {
     }
     
     private void sortOpenedTopComponents() {
+        if (getKind() == Constants.MODE_KIND_EDITOR) {
```
Comment (matthiasblaesing) [inline platform/core.windows/src/org/netbeans/core/windows/model/DefaultModeModel.java]: The if-cascade with early returns makes the function more complex then necessary. I suggest to change it to:

```java
    private void sortOpenedTopComponents() {
        if (getKind() == Constants.MODE_KIND_EDITOR) {
            sortByFile(getSortType());
        } else if( getKind() == Constants.MODE_KIND_SLIDING && Switches.isModeSlidingEnabled()) {
            WindowManagerImpl wm = WindowManagerImpl.getInstance();
            List<TopComponent> opened = topComponentSubModel.getOpenedTopComponents();
            final List<String> prevModes = new ArrayList<String>(opened.size());
            final Map<TopComponent, String> tc2modeName = new HashMap<TopComponent, String>(opened.size());
            for (TopComponent tc : opened) {
                String tcId = wm.findTopComponentID(tc);
                if (null == tcId) {
                    continue;
                }
                ModeImpl prevMode = getTopComponentPreviousMode(tcId);
                if (null == prevMode) {
                    continue;
                }
                if (!prevModes.contains(prevMode.getName())) {
                    prevModes.add(prevMode.getName());
                }
                tc2modeName.put(tc, prevMode.getName());
            }

            if (prevModes.isEmpty()) {
                return; //nothing to sort by (shouldn't really happen)
            }
            Collections.sort(opened, new Comparator<TopComponent>() {
                @Override
                public int compare(TopComponent o1, TopComponent o2) {
                    String mode1 = tc2modeName.get(o1);
                    String mode2 = tc2modeName.get(o2);
                    if (null == mode1 && null != mode2) {
                        return 1;
                    } else if (null != mode1 && null == mode2) {
                        return -1;
                    }
                    return prevModes.indexOf(mode1) - prevModes.indexOf(mode2);
                }
            });
            topComponentSubModel.setOpenedTopComponents(opened);
        }
    }
```

Comment (junichi11) [reply platform/core.windows/src/org/netbeans/core/windows/model/DefaultModeModel.java]: Thank you for looking at this, Matthias!
I'll do it :)

Comment (junichi11) [reply platform/core.windows/src/org/netbeans/core/windows/model/DefaultModeModel.java]: Fixed.

**Comment by matthiasblaesing** [comment]
I have two further comments:
- the package `org.netbeans.core.windows.options` is exported. While friend-only this makes it API, which should be documented. https://github.com/matthiasblaesing/incubator-netbeans/commit/cc3d7d0e4df3f10b0094c6467069d9a253f14d2c
- the preference you added only affects the editor tabs (which is good), this should IMHO be reflected in the option name. https://github.com/matthiasblaesing/incubator-netbeans/commit/3cdd340b8ba13be03b111d417c6f79fd588e906b

**Comment by junichi11** [comment]
Done.

**Comment by geertjanw** [comment]
No objection, thanks, cool feature!

**Pull request merged**